### PR TITLE
Eng 7959 test scalar subquery

### PR DESCRIPTION
**Pull request by johnTcrawford**
Pull request for my sqlcoverage tests of scalar subqueries in the SELECT, FROM, WHERE, ORDER BY, LIMIT and OFFSET clauses.

### CODE REVIEW

**Comment by johnTcrawford** [comment]
I made some improvements: now ready to review, though I still need to fix the sixth pattern (the one with 'A6' in it), to produce queries.

**Comment by johnTcrawford** [comment]
OK, ready to review now (really, I mean it this time).

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<configure-default.sql>
+
+-- DML: purge and regenerate random data first
+DELETE FROM @dmltable
+INSERT INTO @dmltable VALUES (@insertvals)
+
+--- Define "place-holders" used in the queries below
+{_optionalascdesc |= ""}
+{_optionalascdesc |= "ASC"}
+{_optionalascdesc |= "DESC"}
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: _optionalascdesc is just a reinvention of _sortorder which is already included from grammar.sql
For consistency, use _sortorder  -- or, if you much prefer the name, do a global replace of _sortorder w/ _optionalascdesc in all existing scripts.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+{_optionaloderby1limitoffset |= ""}
+{_optionaloderby1limitoffset |= "LIMIT 1000"}
+{_optionaloderby1limitoffset |= "ORDER BY 1 _optionalascdesc _optionallimitoffset"}
+{_optionaloderby2limitoffset |= ""}
+{_optionaloderby2limitoffset |= "LIMIT 1000"}
+{_optionaloderby2limitoffset |= "ORDER BY 1 _optionalascdesc, 2 _optionalascdesc _optionallimitoffset"}
+
+
+-- TEMP, for debugging, just so I can quickly see what data was generated:
+SELECT * FROM @fromtables ORDER BY @idcol
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: Minor: Should this debug query be disabled by default in the checked-in version?

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+SELECT * FROM (SELECT @agg(_variable) FROM @fromtables WHERE _variable[@comparabletype] _cmp _variable[@comparabletype]) AS A4 _optionaloderby1limitoffset
+
+--- Queries with scalar subqueries in the WHERE clause (with optional ORDER BY, LIMIT or OFFSET clauses)
+SELECT _variable AS C0 FROM @fromtables A5 WHERE _variable _cmp (SELECT @agg(_variable) FROM @fromtables                                  ) _optionaloderbylimitoffset
+-- TODO: this produces no queries, unless I remove 'A6.'; I don't understand why:
+SELECT _variable AS C0 FROM @fromtables A6 WHERE _variable _cmp (SELECT @agg(_variable) FROM @fromtables WHERE A6._variable _cmp _variable) _optionaloderbylimitoffset
+
+--- Queries with scalar subqueries in the ORDER BY clause (these currently return errors)
+SELECT (SELECT @agg(_variable) FROM @fromtables                                                                    ) C0 FROM @fromtables A7 ORDER BY C0 _optionalascdesc _optionallimitoffset
+SELECT (SELECT @agg(_variable) FROM @fromtables WHERE _variable[@comparabletype] _cmp A8._variable[@comparabletype]) C0 FROM @fromtables A8 ORDER BY C0 _optionalascdesc _optionallimitoffset
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: Is this the only supported syntax for ordering by a scalar subquery or are we also planning to support:

    select * from R1 ORDER BY (SELECT DESC FROM R2 WHERE R1.ID = ID);

If we decide not to support this syntax, we should make sure that we at least have a junit case that verifies that it gets cleanly rejected -- we probably don't need sqlcoverage permutations to verify the categorical non-support.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+SELECT @idcol AS C0, (SELECT @agg(_variable) FROM @fromtables WHERE A2._variable[@comparabletype] _cmp _variable[@comparabletype]) FROM @fromtables AS A2 _optionaloderbylimitoffset
+
+--- Queries with scalar subqueries in the FROM clause (with optional ORDER BY, LIMIT or OFFSET clauses)
+SELECT * FROM (SELECT @agg(_variable) FROM @fromtables                                                                 )    A3 _optionaloderby1limitoffset
+SELECT * FROM (SELECT @agg(_variable) FROM @fromtables WHERE _variable[@comparabletype] _cmp _variable[@comparabletype]) AS A4 _optionaloderby1limitoffset
+
+--- Queries with scalar subqueries in the WHERE clause (with optional ORDER BY, LIMIT or OFFSET clauses)
+SELECT _variable AS C0 FROM @fromtables A5 WHERE _variable _cmp (SELECT @agg(_variable) FROM @fromtables                                  ) _optionaloderbylimitoffset
+-- TODO: this produces no queries, unless I remove 'A6.'; I don't understand why:
+SELECT _variable AS C0 FROM @fromtables A6 WHERE _variable _cmp (SELECT @agg(_variable) FROM @fromtables WHERE A6._variable _cmp _variable) _optionaloderbylimitoffset
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: The issue here is the number of independent '_variable's being generated without replacement from a smaller number of schema columns.
Removing just the A6. had no effect when I tried it.
The workaround is to correlate two or more by repeating the syntax '_variable[#somelabel]'.
Repeated references can also use the abbreviation "__[#label]" (two underbars) for all but one (typically, the first) reference.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+{_optionallimitoffset |= "LIMIT 4 _optionaloffset"}
+{_optionaloderbylimitoffset |= ""}
+{_optionaloderbylimitoffset |= "LIMIT 1000"}
+{_optionaloderbylimitoffset |= "ORDER BY C0 _optionalascdesc _optionallimitoffset"}
+{_optionaloderby1limitoffset |= ""}
+{_optionaloderby1limitoffset |= "LIMIT 1000"}
+{_optionaloderby1limitoffset |= "ORDER BY 1 _optionalascdesc _optionallimitoffset"}
+{_optionaloderby2limitoffset |= ""}
+{_optionaloderby2limitoffset |= "LIMIT 1000"}
+{_optionaloderby2limitoffset |= "ORDER BY 1 _optionalascdesc, 2 _optionalascdesc _optionallimitoffset"}
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: misspelled order as oder.
_optionaloderby2limitoffset defined but not used

**Comment by paulmartel** [comment]
Finished review round.

**Pull request closed without merging**